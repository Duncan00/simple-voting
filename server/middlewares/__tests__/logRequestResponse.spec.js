'use strict';

const logRequestResponse = require('../logRequestResponse');
const skipRoutes = require('../logRequestResponse/skip_routes');

describe('logRequestResponse', () => {
	const logger = {
		info: () => {}
	};
	beforeEach(() => {
		jest.spyOn(logger, 'info');
	});
	afterEach(() => {
		jest.clearAllMocks();
	});
	test('logRequestResponse', async () => {
		const ctx = {
			logger,
			get: jest.fn(s => s),
			state: {
				request_cycle_trace_id: 'request_cycle_trace_id'
			},
			query: {
				product: 'product'
			},
			path: '/whatever',
			request: {
				body: {
					mock_post_payload: true
				}
			},
			body: 'mock_body',
			status: 'mock_status'
		};
		const next = jest.fn();
		await logRequestResponse()(ctx, next);
		expect(logger.info.mock.calls).toMatchSnapshot('logger.info calls');
		expect(next).toHaveBeenCalledTimes(1);
	});

	test('logRequestResponse with options', async () => {
		const options = {
			extract_ctx_data: jest.fn()
		};
		const ctx = {
			logger,
			get: jest.fn(s => s),
			state: {
				request_cycle_trace_id: 'request_cycle_trace_id'
			},
			query: {
				product: 'product'
			},
			path: '/whatever',
			request: {
				body: {
					mock_post_payload: true
				}
			},
			body: 'mock_body',
			status: 'mock_status'
		};
		const next = jest.fn();
		await logRequestResponse(options)(ctx, next);
		expect(logger.info.mock.calls).toMatchSnapshot(
			'logger.info calls with options'
		);
		expect(next).toHaveBeenCalledTimes(1);
	});

	test('list of skip routes', async () => {
		expect(skipRoutes).toMatchSnapshot();
	});

	skipRoutes.forEach(route => {
		test(`${route} path should be skipped`, async () => {
			const ctx = {
				path: `/${route}`,
				logger
			};
			const next = jest.fn();
			await logRequestResponse()(ctx, next);
			expect(logger.info.mock.calls).toEqual([]);
			expect(next).toHaveBeenCalledTimes(1);
		});
	});
});
