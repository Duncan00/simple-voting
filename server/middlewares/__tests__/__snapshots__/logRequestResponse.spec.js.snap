// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logRequestResponse list of skip routes 1`] = `
Array [
  "whoami",
  "healthz",
]
`;

exports[`logRequestResponse logRequestResponse with options: logger.info calls with options 1`] = `
Array [
  Array [
    Object {
      "message": "accessing path",
      "path": "/whatever",
      "query": Object {
        "product": "product",
      },
      "request": Object {
        "body": Object {
          "mock_post_payload": true,
        },
      },
      "request_body": Object {
        "mock_post_payload": true,
      },
      "request_cycle_trace_id": "request_cycle_trace_id",
      "user_id": "x-authorization-sub",
    },
  ],
  Array [
    Object {
      "message": "returning response",
      "request_cycle_trace_id": "request_cycle_trace_id",
      "response_body": "mock_body",
      "response_status": "mock_status",
    },
  ],
]
`;

exports[`logRequestResponse logRequestResponse: logger.info calls 1`] = `
Array [
  Array [
    Object {
      "message": "accessing path",
      "path": "/whatever",
      "query": Object {
        "product": "product",
      },
      "request": Object {
        "body": Object {
          "mock_post_payload": true,
        },
      },
      "request_body": Object {
        "mock_post_payload": true,
      },
      "request_cycle_trace_id": "request_cycle_trace_id",
      "user_id": "x-authorization-sub",
    },
  ],
  Array [
    Object {
      "message": "returning response",
      "request_cycle_trace_id": "request_cycle_trace_id",
      "response_body": "mock_body",
      "response_status": "mock_status",
    },
  ],
]
`;
